src/vma/sock/sockinfo_udp.cpp

```diff
@@ -913,114 +916,113 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
 				break;
 
 			case IP_ADD_MEMBERSHIP:
-			case IP_DROP_MEMBERSHIP:	
+			case IP_DROP_MEMBERSHIP:
+			case IP_ADD_SOURCE_MEMBERSHIP:
+			case IP_DROP_SOURCE_MEMBERSHIP:
 				{
-				if (!m_sock_offload) {
-					si_udp_logdbg("VMA Rx Offload is Disabled! calling OS setsockopt() for IPPROTO_IP, %s", setsockopt_ip_opt_to_str(__optname));
-					break;
-				}
+					if (!m_sock_offload) {
+						si_udp_logdbg("VMA Rx Offload is Disabled! calling OS setsockopt() for IPPROTO_IP, %s", setsockopt_ip_opt_to_str(__optname));
+						break;
+					}
 
-				if (__optval == NULL || (__optlen < sizeof(struct ip_mreq))) {
-					si_udp_logdbg("IPPROTO_IP, %s; Bad optval! calling OS setsockopt()", setsockopt_ip_opt_to_str(__optname));
-					break;
-				}
+					if ((NULL == __optval) || (__optlen > sizeof(struct ip_mreq_source) || (__optlen < sizeof(struct ip_mreq)))) {
+						si_udp_logdbg("IPPROTO_IP, %s; Bad optval! calling OS setsockopt()", setsockopt_ip_opt_to_str(__optname));
+						break;
+					}
 
-				struct ip_mreq mreq = *(struct ip_mreq*)__optval;
-				if (__optlen >= sizeof(struct ip_mreqn)) {
-					struct ip_mreqn* p_mreqn = (struct ip_mreqn*)__optval;
-					if(p_mreqn->imr_ifindex) {
-						net_dev_lst_t* p_ndv_val_lst = g_p_net_device_table_mgr->get_net_device_val_lst_from_index(p_mreqn->imr_ifindex);
-						net_device_val* p_ndev = NULL;
-						if (p_ndv_val_lst && (p_ndev = dynamic_cast <net_device_val *>(*(p_ndv_val_lst->begin())))) {
-							mreq.imr_interface.s_addr = p_ndev->get_local_addr();
-						} else {
-							struct sockaddr_in src_addr;
-							if (get_ipv4_from_ifindex(p_mreqn->imr_ifindex, &src_addr) == 0) {
-								mreq.imr_interface.s_addr = src_addr.sin_addr.s_addr;
-							} else {
-								si_udp_logdbg("setsockopt(%s) will be passed to OS for handling, can't get address of interface index %d ", setsockopt_ip_opt_to_str(__optname), p_mreqn->imr_ifindex);
-								break;
+					struct ip_mreq_source mreqprm = {{((struct ip_mreq_source*)__optval)->imr_multiaddr.s_addr},{((struct ip_mreq_source*)__optval)->imr_interface.s_addr},{0}};
+					if (((IP_ADD_MEMBERSHIP == __optname) || (IP_DROP_MEMBERSHIP == __optname)) && (__optlen >= sizeof(struct ip_mreqn))) {
```

Comment:
I'm not sure if there is a simple way to separate the cases; but there is a bug inside that if() that I will fix.